llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp

```diff
@@ -1615,6 +1615,61 @@ void GCNSchedStage::revertScheduling() {
   DAG.Regions[RegionIdx] = std::pair(DAG.RegionBegin, DAG.RegionEnd);
 }
 
+bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+                                         SlotIndex OriginalIdx,
+                                         SlotIndex RematIdx) const {
+
+  LiveIntervals *LIS = DAG.LIS;
+  MachineRegisterInfo &MRI = DAG.MRI;
+  OriginalIdx = OriginalIdx.getRegSlot(true);
+  RematIdx = std::max(RematIdx, RematIdx.getRegSlot(true));
+  for (const MachineOperand &MO : InstToRemat->operands()) {
+    if (!MO.isReg() || !MO.getReg() || !MO.readsReg())
+      continue;
+
+    // Do not attempt to reason about PhysRegs
+    if (!MO.getReg().isVirtual()) {
+      assert(DAG.MRI.isConstantPhysReg(MO.getReg()) ||
+             DAG.TII->isIgnorableUse(MO));
```

Comment:
I do not think you can assert it. This can be m0 or flat_scr for example.

Reply:
We also need some type of exec modeling to prevent illegal remats -- otherwise we may remat a use of a more restrictive $exec mask to a less restrictive section. https://godbolt.org/z/qWh47GdWG

Reply:
Actually, after second thought, I think https://godbolt.org/z/qWh47GdWG is not correct control flow.

The case we care about is when: 1. we have a single def, 2. there is a use in a block with a more permissive $exec mask. If we are to remat the def for that use, we will end up using bits which should have been masked out.

However, I don't think such a structure is produceable by control flow. The register will either have multiple defs for separate incoming blocks, or a phi node (in which case we won't be doing remat anyways). PSDB looks good so far, though I plan to do more thorough testing of this (relax conditions for rematerialization)

That said, we may need to disable rematerialization if the kernel has exec handling for WWM / WQM -- looking in to this

Reply:
> This can be m0 or flat_scr for example.

I think this would be covered by `DAG.MRI.isConstantPhysReg(MO.getReg()` ?

Reply:
It is not constant. We assign to m0 all the time, and on some targets initialize flat_scr.

Reply:
> WWM

WQM pass takes care of this, creating a new WWM region for all moved code. However, there seems to be a bug with RA there, so I have this: https://github.com/llvm/llvm-project/pull/123124. This one is pending while we are looking for a better solution.

Reply:
> It is not constant. We assign to m0 all the time, and on some targets initialize flat_scr.

Alright, sure. 

The assert was meant to check that the client of function has are filtered out instructions with bad PhysReg uses (like we do after https://github.com/llvm/llvm-project/commit/e77d428e46d94e1be6e5f38205b01d3f528d5e3f ) but we can just give up on them.

Reply:
> I have this: https://github.com/llvm/llvm-project/pull/123124. This one is pending while we are looking for a better solution.

I wonder if this PR should be blocked for that work?

Reply:
> > I have this: #123124. This one is pending while we are looking for a better solution.
> 
> I wonder if this PR should be blocked for that work?

No, not at all. In general WQM should just handle it.